### DISCUSSION TITLE
Time limit resets, when respawning on a checkpoint

### DISCUSSION THREAD

**Discussion by provigz** [SuperTux/supertux#2060] [General]
**SuperTux version:** 0.6.3
**System information:** Windows 10

##### Expected behavior
Time limit does not reset, when respawning on a checkpoint.
##### Actual behavior
Time limit resets, when respawning on a checkpoint.
##### Steps to reproduce actual behavior
Enter a level with a time limit, activate a checkpoint and die. Notice that the time limit resets.

**Comment by Semphriss** [comment]
### Proposed solution:
**Save the time at the moment of taking the checkpoint and restore that time when respawning at the checkpoint.**

**Reply by Semphriss** [reply]
#### Problem:

If the player takes a checkpoint with not enough time left (let's say 3 seconds), they will be forced to die in loop at that checkpoint without any chance of winning, causing a softlock.

**Comment by Semphriss** [comment]
### Proposed solution:
**Retain the current behavior: Timers reset to their initial time when respawning at a checkpoint**

**Reply by Semphriss** [reply]
#### Problem:

This defeats the point of timers and encourages players to pointlessly die to get more time.

**Reply by Semphriss** [reply]
#### Support:

This gives a merciful advantage to an already disadvantaged player.

**Comment by Semphriss** [comment]
### Proposed solution:

**Do not change the time left in anyway when respawning. Keep the timers running through deaths; retain the remaining time at the moment of death and restore that same time upon respawning at a checkpoint. The time in a timer never goes up while playing the level, even at deaths.**

**Reply by Semphriss** [reply]
#### Support:

This is intuitive enough, allows retaining the purpose of timers, and does not require special thought from level designers.

**Reply by Semphriss** [reply]
#### Problem:

This allows the possibility of respawning at a checkpoint with not enough time left to finish the level. But this, IMO, does not contribute to a poor gameplay experience, as players are already used to dying from not having enough time left even without checkpoints.

**Reply by provigz** [reply]
I like this solution, and I recommend it works in the following way in detail:

* Time limit continues counting, if the player dies from something else and respawns on a checkpoint.
* Time limit resets, if the player dies and goes back to the beginning (has no checkpoints).
* Time limit resets and the player goes back to the beginning, when dying from time limit. (Even if they have active checkpoints.)

**Comment by Alasdairbugs** [comment]
My proposed solution:

upon death, take the original level time, divide it by the number of checkpoints (+ 1) in the level, and subtract the time from the original amount depending on what checkpoint you are on.

example: time is 300s. 1 checkpoint in the level. 1+1 is 2, 300/2 is 150. upon respawning, you get 150s of time limit. This assumes the level maker puts the checkpoints at relative equidistance to the start, end, and other checkpoints - which is what level designers should be doing anyway.

**Reply by Alasdairbugs** [reply]
also, say there is a time limit of 300s in a level with 2 checkpoints. 2+1 is 3, 300/3 is 100, but lets say you respawn at the SECOND checkpoint, and not the first. you would have 100s left. if you respawn at the first checkpoint, you have 200s left.

**Reply by Semphriss** [reply]
#### Problem:

If the checkpoints aren't spread out evenly, or if the level isn't linear, this might create cases of softlock.

It's possible to make well-designed levels with that technique, but many levels are known to be particularly poorly designed, so it should be taken into account.

**Reply by Semphriss** [reply]
Also, how would the game know which checkpoint is which in the ordering? We can't rely on simple left-to-right ordering, it won't work for levels that aren't a straight line; we can't rely on the order in which the player hits the checkpoints, because people can skip checkpoints, or pass to a checkpoint several times.

It might also be difficult to track checkpoints across sectors, since we can't evaluate how to order sectors by progress.Sectors also rarely are linked end-to-end in a level's continuity, often you have a sector that you can access through a door that's in the middle of another sector.

Additionally, if he level branches out to an optional bonus zone, it's not clear how it should be handled. Should it "rewind" some time to let the player complete the bonus zone?


-----------------------


(Apologies if it looks like I'm spamming problems, I just want to have those dealt with as early as possible to avoid having to deal with them at a point where plenty of levels we didn't keep track of are having problems of all kinds)

**Reply by provigz** [reply]
A possible way this could be handled is to have the level designer set the time, which the player will have left, when they respawn from a specific checkpoint. The problem is it requires thought for level designers + changes to all current levels, that have a time limit, so it's really not a good way to handle it.

**Reply by Semphriss** [reply]
Plus there's the case: In case level designers forget/don't bother to set that value, what should it default to? We can't really leave an errorful state like "0" since it's to be expected that testers don't respawn at each checkpoint...

That's in addition to having either a value in the editor that's useless for levels without time limits, or having to add a check if the level has a time limit upon loading and saving settings, etc.
